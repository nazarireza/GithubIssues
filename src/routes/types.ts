import { RouteProp } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';

export enum Routes {
  Configuration = 'configuration',
  Issues = 'issues',
  IssueDetail = 'issue_detail',
  BookmarkedIssues = 'bookmarked_issues',
}

export type RootStackParamList = {
  [Routes.Configuration]: undefined;
  [Routes.Issues]: undefined;
  [Routes.BookmarkedIssues]: undefined;
  [Routes.IssueDetail]: {
    issueNumber: number;
  };
};

export type RootStackComponent<RouteName extends keyof RootStackParamList> =
  React.FC<{
    navigation: NativeStackNavigationProp<RootStackParamList, RouteName>;
    route: RouteProp<RootStackParamList, RouteName>;
  }>;
