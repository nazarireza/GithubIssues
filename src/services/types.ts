export type IssueState = 'open' | 'closed' | 'all';

export type IssueDto = {
  number: number;
  title: string;
  body: string;
  state: IssueState;
  created_at: string;
};

export type GetIssueModel = {
  organization: string;
  repository: string;
  page?: number;
  per_page?: number;
  state?: IssueState;
};

export type GetSingleIssueModel = {
  organization: string;
  repository: string;
  issueNumber: number;
};

export type RepositoryDto = {
  id: number;
  name: string;
};

export type GetRepositoryModel = {
  organization: string;
  repository: string;
};
