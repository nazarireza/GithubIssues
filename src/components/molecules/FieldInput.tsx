import React, { forwardRef, memo } from 'react';
import { TextInputProps, View, StyleSheet, TextInput } from 'react-native';
import { Input } from '../atoms/Input';
import { Label } from '../atoms/Label';
import { Space } from '../atoms/Space';

type FieldInputProps = {
  title?: string;
  hasError?: boolean;
  errorMessages?: string[];
};

export const FieldInput = memo(
  forwardRef<TextInput, FieldInputProps & TextInputProps>(
    ({ title, hasError, errorMessages, ...rest }, ref) => {
      return (
        <View>
          <Label>{title}</Label>
          <Space />
          <Input testID="input" ref={ref} {...rest} />
          <Space />
          {hasError && (
            <Label style={styles.errorMessage}>{errorMessages?.[0]}</Label>
          )}
        </View>
      );
    }
  )
);

const styles = StyleSheet.create({
  errorMessage: {
    color: 'red',
  },
});
