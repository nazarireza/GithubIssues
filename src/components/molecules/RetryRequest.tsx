import { memo } from 'react';
import { View, StyleSheet } from 'react-native';
import { useTranslation } from '../../assets/dictionary';
import { Button } from '../atoms/Button';
import { Label } from '../atoms/Label';
import { Space } from '../atoms/Space';

type RetryRequestProps = {
  message?: string;
  onRetry?: () => void;
};

export const RetryRequest: React.FC<RetryRequestProps> = memo(
  ({ message, onRetry }) => {
    const { t } = useTranslation();

    message = message || t("Something's Wrong");
    return (
      <View style={styles.container}>
        <Label>{message}</Label>
        <Space />
        <Button title={t('Retry')} onPress={onRetry} />
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
  },
});
