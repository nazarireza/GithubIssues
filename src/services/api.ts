import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import Constants from 'expo-constants';
import {
  GetIssueModel,
  GetRepositoryModel,
  GetSingleIssueModel,
  IssueDto,
  RepositoryDto,
} from './types';

export const api = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: Constants.manifest?.extra?.baseUrl,
  }),

  endpoints: (build) => ({
    getRepoDetail: build.query<RepositoryDto[], GetRepositoryModel>({
      query: ({ organization, repository }) => ({
        url: `/repos/${organization}/${repository}`,
      }),
    }),

    getIssues: build.query<IssueDto[], GetIssueModel>({
      query: ({
        state = 'open',
        page = 0,
        per_page = 10,
        organization,
        repository,
      }) => ({
        url: `/repos/${organization}/${repository}/issues?state=${state}&page=${page}&per_page=${per_page}`,
      }),
    }),

    getSingleIssue: build.query<IssueDto, GetSingleIssueModel>({
      query: ({ organization, repository, issueNumber }) => ({
        url: `/repos/${organization}/${repository}/issues/${issueNumber}`,
      }),
    }),
  }),
});

export const {
  useLazyGetIssuesQuery,
  useGetSingleIssueQuery,
  useLazyGetRepoDetailQuery,
} = api;
