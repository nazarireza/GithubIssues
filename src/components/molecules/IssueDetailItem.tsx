import { memo } from 'react';
import { StyleSheet, View } from 'react-native';
import { IssueDto } from '../../services/types';
import { Label } from '../atoms/Label';
import { Space } from '../atoms/Space';
import Markdown from 'react-native-markdown-display';
import { IconButton } from '../atoms/IconButton';

type IssueDetailItemProps = {
  item: IssueDto;
  isBookmarked?: boolean;
  onToggleBookmark?: () => void;
};

export const IssueDetailItem: React.FC<IssueDetailItemProps> = memo(
  ({ item, isBookmarked = false, onToggleBookmark }) => {
    return (
      <View>
        <Label style={styles.title}>{item.title}</Label>
        <Space />
        <IconButton
          testID="bookmark"
          name={isBookmarked ? 'star' : 'star-outline'}
          onPress={onToggleBookmark}
        />
        <Space />
        <Markdown>{item.body}</Markdown>
      </View>
    );
  }
);

const styles = StyleSheet.create({
  title: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  headerContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});
