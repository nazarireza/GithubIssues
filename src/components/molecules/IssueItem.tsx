import { memo } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  TouchableOpacityProps,
} from 'react-native';
import { IssueDto } from '../../services/types';
import { Card } from '../atoms/Card';
import { Label } from '../atoms/Label';

type IssueItemProps = {
  item: IssueDto;
  isBookmarked?: boolean;
};

export const IssueItem: React.FC<IssueItemProps & TouchableOpacityProps> = memo(
  ({
    item: { id, title, body, state, created_at },
    isBookmarked = false,
    ...rest
  }) => {
    return (
      <TouchableOpacity activeOpacity={0.7} {...rest}>
        <Card>
          <Label style={styles.title}>{title}</Label>
        </Card>
      </TouchableOpacity>
    );
  }
);

const styles = StyleSheet.create({
  title: {
    fontSize: 16,
  },
});
