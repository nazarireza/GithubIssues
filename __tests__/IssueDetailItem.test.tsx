import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from 'react-native-testing-library';
import { IssueDetailItem } from '../src/components/molecules/IssueDetailItem';
import { IssueDto } from '../src/services/types';

const item: IssueDto = {
  number: 1234,
  title: 'title1',
  body: '##body1',
  created_at: '2020-01-01',
  state: 'open',
};

describe('<IssueDetailItem />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<IssueDetailItem item={item} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('fire the onToggleBookmark', () => {
    const mockOnToggleBookmark = jest.fn();
    const { getByTestId } = render(
      <IssueDetailItem item={item} onToggleBookmark={mockOnToggleBookmark} />
    );

    const button = getByTestId('bookmark');
    fireEvent(button, 'onPress');

    expect(mockOnToggleBookmark).toHaveBeenCalledTimes(1);
  });
});
