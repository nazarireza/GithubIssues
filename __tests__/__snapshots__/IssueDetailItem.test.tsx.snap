// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IssueDetailItem /> renders correctly 1`] = `
<View>
  <Text
    style={
      Object {
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    title1
  </Text>
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "rgba(0,0,0,.1)",
        "borderRadius": 5,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 8,
      }
    }
    testID="bookmark"
  >
    <Icon
      name="star-outline"
      size={30}
    />
  </View>
  <View
    style={
      Object {
        "height": 8,
        "width": 8,
      }
    }
  />
  <View
    style={Object {}}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
          "marginBottom": 10,
          "marginTop": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={Object {}}
      >
        <Text
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          ##body1
        </Text>
      </Text>
    </View>
  </View>
</View>
`;
